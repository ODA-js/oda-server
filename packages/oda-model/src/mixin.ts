import {
  EntityBaseMetaInfo,
  EntityBaseInput,
  EntityBasePersistence,
  IEntityBase,
  EntityBase,
  EntityBaseInternal,
} from './entitybase';
import { merge } from 'lodash';

export interface IMixin extends IEntityBase<MixinMetaInfo, MixinPersistence> {}
export interface MixinPersistence extends EntityBasePersistence {}
export interface MixinInput
  extends EntityBaseInput<MixinMetaInfo, MixinPersistence> {}
export interface MixinInternal
  extends EntityBaseInternal<MixinMetaInfo, MixinPersistence> {}
export interface MixinMetaInfo extends EntityBaseMetaInfo<MixinPersistence> {}

const defaultMetaInfo = {};
const defaultInternal = {};
const defaultInput = {};

export class Mixin
  extends EntityBase<MixinMetaInfo, MixinInput, MixinInternal, MixinPersistence>
  implements IMixin {
  constructor(inp: MixinInput) {
    super(merge({}, defaultInput, inp));
    this.metadata_ = merge({}, defaultMetaInfo, this.metadata_);
    this.$obj = merge({}, defaultInternal, this.$obj);
  }

  public toObject(): MixinInput {
    return merge({}, super.toObject(), {} as Partial<MixinInput>);
  }
}
