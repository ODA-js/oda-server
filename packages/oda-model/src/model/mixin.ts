import { MetaModelType, IMixin, MixinStorage, MixinInput } from './interfaces';
import { EntityBase } from './entitybase';

/**
 * 1. тип объекта который входит на updateWith
 * 2. тип объекта который идет на toObject
 * 3. тип объекта который идет на toJSON
 * 3. тип объекта который идет на выходе clone
 */

export class Mixin extends EntityBase implements IMixin {
  public modelType: MetaModelType = 'mixin';
  protected $obj!: MixinStorage;

  constructor(obj: MixinInput) {
    super(obj);
  }
}
