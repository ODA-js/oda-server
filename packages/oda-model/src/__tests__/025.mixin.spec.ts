import 'jest';
import { Mixin } from '../mixin';

describe('Mixin', () => {
  it('default', () => {
    const res = new Mixin({ name: 'IMix  ', fields: [{ name: 'id' }] });
    expect(res.toObject()).toMatchSnapshot('default');
  });
});
