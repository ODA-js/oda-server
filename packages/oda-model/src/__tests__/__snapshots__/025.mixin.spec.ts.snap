// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mixin default: default 1`] = `
Object {
  "description": undefined,
  "fields": Array [
    Object {
      "args": Object {},
      "description": undefined,
      "inheritedFrom": undefined,
      "metadata": Object {
        "acl": Object {
          "read": Array [],
          "update": Array [],
        },
        "entity": "IMix",
        "order": 0,
        "persistence": Object {
          "derived": false,
          "identity": true,
          "indexed": true,
          "persistent": true,
          "required": true,
        },
      },
      "multiplicity": "one",
      "name": "id",
      "title": undefined,
      "type": "String",
    },
  ],
  "metadata": Object {
    "UI": Object {
      "edit": Array [],
      "embedded": Array [],
      "hidden": Array [],
      "list": Array [],
      "listName": Array [],
      "quickSearch": Array [],
      "show": Array [],
    },
    "acl": Object {
      "create": Array [],
      "delete": Array [],
      "readMany": Array [],
      "readOne": Array [],
      "update": Array [],
    },
    "name": Object {
      "plural": "IMixes",
      "singular": "IMix",
    },
    "persistence": Object {
      "indexes": Object {
        "id": Object {
          "fields": Object {
            "id": 1,
          },
          "name": "id",
          "options": Object {
            "unique": true,
          },
        },
      },
    },
    "titlePlural": "IMixes",
  },
  "name": "IMix",
  "operations": Array [],
  "title": undefined,
}
`;
